.mage/proto.go

```diff
@@ -0,0 +1,215 @@
+//go:build mage
+// +build mage
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+	"go.einride.tech/mage-tools/mglog"
+	"go.einride.tech/mage-tools/mgtool"
+	"go.einride.tech/mage-tools/targets/mgbuf"
+)
+
+type Proto mg.Namespace
+
+func (Proto) All() {
+	mg.SerialDeps(
+		mg.F(Proto.ClangFormat, "einride"),
+		Proto.BufLint,
+		Proto.ApiLinterLint,
+		Proto.BufGenerate,
+	)
+}
+
+func (Proto) BufLint() {
+	mg.Deps(mgbuf.BufLint)
+}
+
+func (Proto) ProtocGenGo() error {
+	logger := mglog.Logger("protoc-gen-go")
+	newer, err := isDependenciesNewer("build/protoc-gen-go", []string{"../go.mod"})
+	if err != nil {
+		return nil
+	}
+	if !newer {
+		return nil
+	}
+	logger.Info("building binary...")
+	return sh.Run("go", "build", "-o", "build/protoc-gen-go", "google.golang.org/protobuf/cmd/protoc-gen-go")
+}
+
+func (Proto) ProtocGenGoAip() error {
+	logger := mglog.Logger("protoc-gen-go-aip")
+	logger.Info("building binary...")
+	return sh.Run("go", "build", "-o", "build/protoc-gen-go-aip", "../cmd/protoc-gen-go-aip")
+}
+
+func (Proto) BufGenerate(ctx context.Context) error {
+	logger := mglog.Logger("buf-generate")
+	ctx = logr.NewContext(ctx, logger)
+	goGrpc, err := mgtool.GoInstall(ctx, "google.golang.org/grpc/cmd/protoc-gen-go-grpc", "v1.2.0")
+	if err != nil {
+		return err
+	}
+	mg.Deps(
+		Proto.ProtocGenGo,
+		Proto.ProtocGenGoAip,
+	)
+	path := fmt.Sprintf("%s:%s:%s", os.Getenv("PATH"), filepath.Dir(goGrpc), "../.mage/tools/protoc/3.15.7/bin")
+	if err := os.Setenv("PATH", path); err != nil {
+		return err
+	}
+	logger.Info("generating protobuf stubs...")
+	return mgbuf.Buf(ctx, []string{"generate", "--template", "buf.gen.yaml", "--path", "einride"}...)
+}
+
+func (Proto) BuildDescriptorPB(ctx context.Context) error {
+	logger := mglog.Logger("build/descriptor.pb")
+	ctx = logr.NewContext(ctx, logger)
+	var protoFiles []string
+	err := filepath.WalkDir("einride", func(path string, d fs.DirEntry, err error) error {
+		if filepath.Ext(path) == ".proto" {
+			protoFiles = append(protoFiles, path)
+		}
+		return nil
+	})
```

Comment:
nit: might be worthwhile to extract this to a function `protoFilesIn(dir string) ([]string, error)` since it is done in multiple places in this file. I wish `filepath.Glob` would support this out of the box, but no..

Reply:
I was thinking about this while i wrote this a few times, not sure why i didnt :P